Add tests of creating snapshots with lock_managers

qiankehan opened the pull request:
Test for https://bugzilla.redhat.com/show_bug.cgi?id=1191901
Check deadlock when create internal or external snapshot under
lock_managers.

qiankehan commented:
Test result:
sudo avocado run --vt-type libvirt type_specific.io-github-autotest-libvirt.lock_manager
JOB ID     : 397a3cff87b67089416ff02df1ba307abbda1eff
JOB LOG    : /root/avocado/job-results/job-2017-04-14T09.22-397a3cf/job.log
TESTS      : 4
 (1/4) type_specific.io-github-autotest-libvirt.lock_manager.internal_snapshot.virtlockd: PASS (37.38 s)
 (2/4) type_specific.io-github-autotest-libvirt.lock_manager.internal_snapshot.sanlock: PASS (34.69 s)
 (3/4) type_specific.io-github-autotest-libvirt.lock_manager.external_snapshot.virtlockd: PASS (35.62 s)
 (4/4) type_specific.io-github-autotest-libvirt.lock_manager.external_snapshot.sanlock: PASS (43.80 s)
RESULTS    : PASS 4 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0
JOB HTML   : /root/avocado/job-results/job-2017-04-14T09.22-397a3cf/html/results.html
TESTS TIME : 151.49 s

balamuruhans commented on libvirt/tests/src/lock_manager.py:
```diff
@@ -0,0 +1,113 @@
+        """
+        augtool = augeas.Augeas()
+        for key in conf_dict:
+            aug_path = '/files' + file_path + '/' + key
```
please use `os.path.join()` to append paths

qiankehan replied:
Thks for review. All path joining has been updated.

balamuruhans commented on libvirt/tests/src/lock_manager.py:
```diff
@@ -0,0 +1,113 @@
+        """
+        augtool = augeas.Augeas()
+        for key in conf_dict:
+            aug_path = '/files' + file_path + '/' + key
```
please use `os.path.join()` to append paths

balamuruhans commented on libvirt/tests/src/lock_manager.py:
```diff
@@ -0,0 +1,113 @@
+                    "host_id":1,
+                    "disk_lease_dir":"/var/lib/libvirt/sanlock",
+                    "user":"sanlock",
+                    "group":"sanlock"}
```
just curious that should the `virtlockd_conf` and `sanlock_conf` attributes hardcoded or can be made configurable from params

balamuruhans commented on libvirt/tests/src/lock_manager.py:
```diff
@@ -0,0 +1,113 @@
+            snap_name + " " + snap_create_option,
+            timeout=snap_timeout)
+        logging.debug(create_snap_result)
+        if create_snap_result.stderr is not '':
```
`libvirt.check_exit_status(create_snap_result)` can be used

qiankehan replied:
libvirt.check_exit_status has been implemented for error check.

balamuruhans commented on libvirt/tests/src/lock_manager.py:
```diff
@@ -0,0 +1,113 @@
+import logging
+import augeas
+
+from autotest.client.shared import error
```
`test.fail()`, `test.error()` and `test.skip()` can be used instead of `error`

balamuruhans commented on libvirt/tests/src/lock_manager.py:
```diff
@@ -0,0 +1,113 @@
+            vm.destroy()
+        rm_conf("/etc/libvirt/qemu.conf", qemu_conf)
+        snap_metadata_path = "/var/lib/libvirt/qemu/snapshot/" +\
+            vm_name + "/" + snap_name + ".xml"
```
same here

balamuruhans commented on libvirt/tests/src/lock_manager.py:
```diff
@@ -0,0 +1,113 @@
+    snap_delete_option = params.get("delete_option", "")
+    snap_timeout = params.get("snap_timeout")
+    lock_type = params.get("lock_type")
+    qemu_conf = {"lock_manager":lock_type}
```
space after colon in dict here and elsewhere will fix travis check

balamuruhans commented on libvirt/tests/src/lock_manager.py:
```diff
@@ -0,0 +1,111 @@
+            augtool.set(aug_path, conf_dict[key])
+        augtool.save()
+
+    def rm_conf(file_path, conf_dict):
```
Both `set_conf` and `rm_conf` are similar, you can have single function,
```
def configure (file_path, conf_dict, func=augtool.save):
    augtool = augeas.Augeas()
    for key in conf_dict:
        aug_path = os.path.join('/files', file_path, key)
        func()
    augtool.save()
```

balamuruhans commented on libvirt/tests/src/lock_manager.py:
```diff
@@ -0,0 +1,111 @@
+                    "host_id": 1,
+                    "disk_lease_dir": "/var/lib/libvirt/sanlock",
+                    "user": "sanlock",
+                    "group": "sanlock"}
```
does it good to hardcode these values

balamuruhans commented on libvirt/tests/src/lock_manager.py:
```diff
@@ -0,0 +1,111 @@
+        backup_xml.sync()
+
+        if lock_type is "virtlockd":
+            rm_conf("/etc/libvirt/qemu-lockd.conf", virtlockd_conf)
```
can this this path be from param variable so that it can be used when using more than once

balamuruhans commented on libvirt/tests/src/lock_manager.py:
```diff
@@ -0,0 +1,111 @@
+            service_mgr.restart("libvirtd")
+            service_mgr.stop("virtlockd")
+        if lock_type is "sanlock":
+            rm_conf("/etc/libvirt/qemu-sanlock.conf", sanlock_conf)
```
same here

balamuruhans commented on libvirt/tests/src/lock_manager.py:
```diff
@@ -0,0 +1,111 @@
+            service_mgr.stop("virtlockd")
+        if lock_type is "sanlock":
+            rm_conf("/etc/libvirt/qemu-sanlock.conf", sanlock_conf)
+            service_mgr.restart("libvirtd")
```
same here, restart libvirtd in clean up is irrespective of sanlock or virtlockd testcase. so can be placed outside the check

balamuruhans commented on libvirt/tests/src/lock_manager.py:
```diff
@@ -0,0 +1,111 @@
+        set_conf("/etc/libvirt/qemu-sanlock.conf", sanlock_conf)
+        service_mgr.start("wdmd")
+        service_mgr.start("sanlock")
+        service_mgr.restart("libvirtd")
```
restarting libvirtd is irrespective of virtlockd or sanlock, can be placed outside the check

will-Do commented:
@qiankehan 
Please response to the comments

qiankehan commented:
@will-Do  @balamuruhans  Sorry for later reply.
I have refactored the function of 'set' or 'remove' configurations and removed the hard-codes.  Please review this new commit. Thanks.

qiankehan commented:
Test result:
```
avocado run --vt-type libvirt lock_manager                                                                                                   
JOB ID     : 12c067bc7d217840dd447706c0c65c0a6e33c1c4
JOB LOG    : /root/avocado/job-results/job-2017-07-07T16.24-12c067b/job.log
 (1/4) type_specific.io-github-autotest-libvirt.lock_manager.internal_snapshot.virtlockd: /** Message: pygobject_register_sinkfunc is deprecated (GstObject)
PASS (60.89 s)
 (2/4) type_specific.io-github-autotest-libvirt.lock_manager.internal_snapshot.sanlock: \** Message: pygobject_register_sinkfunc is deprecated (GstObject)
PASS (57.20 s)
 (3/4) type_specific.io-github-autotest-libvirt.lock_manager.external_snapshot.virtlockd: /** Message: pygobject_register_sinkfunc is deprecated (GstObject)
PASS (60.62 s)
 (4/4) type_specific.io-github-autotest-libvirt.lock_manager.external_snapshot.sanlock: /** Message: pygobject_register_sinkfunc is deprecated (GstObject)
PASS (57.49 s)
RESULTS    : PASS 4 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 248.44 s
```

iccaszhulili commented:
rh-caselink retest please

qiankehan commented:
Closed this PR since qemu enabled OFD lock from v2.10.0. And it is no sense for lock manager.
